### PR TITLE
Search Domains and DNS Resolver on Windows

### LINKED ISSUE #12712: Hosts with DNS suffixes cannot be resolved in Windows 10 and Java 9+
I am using Spring WebClient based on netty and it cannot resolve hosts with DNS suffixes on Windows and Java 9+. I think this [hack](https://github.com/netty/netty/blob/9cd94547d0211c04b610878e5e267c8de2342b97/resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java#L189:L202) is responsible for such behaviour.

Why is reflection used here? You can acquire instance of ResolverConfiguration via static method, no?

Java 8 works fine though. I think issue is simillar to [this](https://github.com/netty/netty/issues/11885).

### Expected behavior
Hosts with dns suffixes resolves succesfully

### Actual behavior
UnknownHostException
```
Caused by: java.net.UnknownHostException: failed to resolve 'my-host' after 8 queries 
	at io.netty.resolver.dns.DnsResolveContext.finishResolve(DnsResolveContext.java:1046)
	at io.netty.resolver.dns.DnsResolveContext.tryToFinishResolve(DnsResolveContext.java:999)
	at io.netty.resolver.dns.DnsResolveContext.query(DnsResolveContext.java:417)
	at io.netty.resolver.dns.DnsResolveContext.onResponse(DnsResolveContext.java:628)
	at io.netty.resolver.dns.DnsResolveContext.access$400(DnsResolveContext.java:65)
	at io.netty.resolver.dns.DnsResolveContext$2.operationComplete(DnsResolveContext.java:461)
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:578)
	at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:571)
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:550)
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:491)
	at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:616)
	at io.netty.util.concurrent.DefaultPromise.setSuccess0(DefaultPromise.java:605)
	at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:104)
	at io.netty.resolver.dns.DnsQueryContext.trySuccess(DnsQueryContext.java:201)
	at io.netty.resolver.dns.DnsQueryContext.finish(DnsQueryContext.java:193)
	at io.netty.resolver.dns.DnsNameResolver$DnsResponseHandler.channelRead(DnsNameResolver.java:1301)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioMessageChannel$NioMessageUnsafe.read(AbstractNioMessageChannel.java:97)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

### Steps to reproduce
Configure DNS suffixes on Windows. This can be achieved like this: edit SearchList property at registry path HKLM\SYSTEM\CurrentControlSet\services\Tcpip\Parameters.

### Netty version
4.1.77.Final
### JVM version (e.g. `java -version`)
```
openjdk version "17.0.2" 2022-01-18
OpenJDK Runtime Environment (build 17.0.2+8-86)
OpenJDK 64-Bit Server VM (build 17.0.2+8-86, mixed mode, sharing)
```
### OS version (e.g. `uname -a`)
Windows 10

### PR DESCRIPTION
**Pull request by flx5**
Motivation:
On Windows with JDK 9+ the search domain and dns servers are missing.

Modification:

This PR uses JNI to fetch the DNS Servers from the operating system.

Result:

Fixes #12712 #11885.

### CODE REVIEW

**Comment by hyperxpro** [comment]
More JNI 😐

File: resolver-dns-native-windows/pom.xml [hunk 1]
```diff
@@ -0,0 +1,148 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Copyright 2019 The Netty Project
```
Comment (normanmaurer) [inline resolver-dns-native-windows/pom.xml:3]:
```suggestion
  ~ Copyright 2022 The Netty Project
```

File: resolver-dns-native-windows/pom.xml [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+  ~ License for the specific language governing permissions and limitations
+  ~ under the License.
+  -->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/maven-v4_0_0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <groupId>io.netty</groupId>
+    <artifactId>netty-parent</artifactId>
+    <version>4.1.85.Final-SNAPSHOT</version>
```
Comment (normanmaurer) [inline resolver-dns-native-windows/pom.xml:23]:
```suggestion
    <version>4.1.86.Final-SNAPSHOT</version>
```

File: resolver-dns-native-windows/pom.xml [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+          <family>windows</family>
+        </os>
+      </activation>
+      <properties>
+        <skipTests>false</skipTests>
+      </properties>
+      <build>
+        <plugins>
+          <plugin>
+            <groupId>com.googlecode.cmake-maven-project</groupId>
```
Comment (normanmaurer) [inline resolver-dns-native-windows/pom.xml:44]: Can we use the hawtjini plugin so we are consistent with the rest of our native stuff ?

File: resolver-dns-native-windows/src/main/c/src/Connection.cpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
+#include "connection.h"
```
Comment (normanmaurer) [inline resolver-dns-native-windows/src/main/c/src/Connection.cpp:1]: missing license header

File: resolver-dns-native-windows/src/main/c/src/Connection.cpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+#include <stdexcept>
+#include <iostream>
+
+#pragma comment(lib, "ws2_32.lib")
+
+Connection::Connection(LPSOCKADDR sock_address) {
+	switch (sock_address->sa_family) {
```
Comment (normanmaurer) [inline resolver-dns-native-windows/src/main/c/src/Connection.cpp:8]: we use 4 spaces... Please update

File: resolver-dns-native-windows/src/main/c/src/JavaInterface.cpp [hunk 6]
```diff
@@ -0,0 +1,63 @@
+#include "JavaInterface.h"
+#include "NetworkAdapters.h"
+#include "NativeExceptionClass.h"
+#include <memory>
+#include "JavaNetworkAdapterClass.h"
+#include <stdexcept>
+
+#define NETTY_JNI_UTIL_JNI_VERSION JNI_VERSION_1_6
```
Comment (normanmaurer) [inline resolver-dns-native-windows/src/main/c/src/JavaInterface.cpp:8]: Can we just use c ? We dont use any c++ atm and I would like to keep it like this to make things easier for the maintainers .

File: resolver-dns/src/main/java/io/netty/resolver/dns/WindowsResolverDnsServerAddressStreamProvider.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+ * under the License.
+ */
+package io.netty.resolver.dns;
+
+import io.netty.resolver.dns.windows.NetworkAdapter;
+import io.netty.resolver.dns.windows.WindowsAdapterInfo;
+import io.netty.util.internal.StringUtil;
+
+import java.net.InetSocketAddress;
+import java.util.*;
```
Comment (normanmaurer) [inline resolver-dns/src/main/java/io/netty/resolver/dns/WindowsResolverDnsServerAddressStreamProvider.java:23]: we dont use wildcard imports

File: resolver-dns-native-windows/src/main/c/src/JavaInterface.cpp (hunk 6 above)
Comment (flx5) [reply resolver-dns-native-windows/src/main/c/src/JavaInterface.cpp]: Ported the native code to c.

File: resolver-dns-native-windows/pom.xml (hunk 3 above)
Comment (flx5) [reply resolver-dns-native-windows/pom.xml:44]: Will look into it, but that plugin had some trouble with newer Visual Studio Editions.

File: resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+		port = ntohs(addr_in->sin6_port);
+		host = (*env)->NewStringUTF(env, ip);
+		break;
+	}
+	default:
+		throwNativeException(env, "Unknown address family %d", sockAddr->sa_family);
+		return NULL;
+	}
+
+    return (*env)->NewObject(env, inetSocketAddressClass, inetSocketAddressCtor, host, port);
```
Comment (normanmaurer) [inline resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c:154]: fix formatting

File: resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+
+    if (loadClass(env, "io/netty/resolver/dns/windows/NetworkAdapter", &networkAdapterClass) != JNI_OK) {
+        return JNI_ERR;
+    }
+
+    if (loadMethod(env, networkAdapterClass, "<init>", "(Ljava/util/List;Ljava/util/List;)V", &networkAdapterCtor) != JNI_OK) {
+        return JNI_ERR;
+    }
+
+    if (loadClass(env, "java/util/LinkedList", &linkedListClass) != JNI_OK) {
```
Comment (normanmaurer) [inline resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c:339]: Use an `ArrayList`... `LinkedList` shouldn't be used 99% of times.

File: resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+        return NULL;
+    }
+
+    jstring result = (*env)->NewStringUTF(env, buffer);
+    free(buffer);
+
+    return result;
+}
+
+jobject searchDomainsToJava(JNIEnv* env, IP_ADAPTER_ADDRESSES* adapter) {
```
Comment (normanmaurer) [inline resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c:204]: Same comment as above. Consider just returning Object[].

File: resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.h [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+ * under the License.
+ */
+
+/* DO NOT EDIT THIS FILE - it is machine generated */
+#include <jni.h>
+/* Header for class io_netty_resolver_dns_windows_NativeWrapper */
+
+#ifndef _Included_io_netty_resolver_dns_windows_NativeWrapper
+#define _Included_io_netty_resolver_dns_windows_NativeWrapper
+#ifdef __cplusplus
```
Comment (normanmaurer) [inline resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.h:23]: as we dont use c++ anymore I think this cane be removed.

File: resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c [hunk 12]
```diff
@@ -0,0 +1,380 @@
...
+
+    free(stringClass);
+
+    free(linkedListAdd);
+    free(linkedListCtor);
+    free(linkedListClass);
+
+    free(inetSocketAddressCtor);
+    free(inetSocketAddressClass);
+}
```
Comment (normanmaurer) [inline resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c:380]: missing EOL

File: pom.xml [hunk 13]
```diff
@@ -1824,8 +1825,8 @@
         </plugin>
         <plugin>
           <groupId>org.fusesource.hawtjni</groupId>
-          <artifactId>maven-hawtjni-plugin</artifactId>
-          <version>1.14</version>
+          <artifactId>hawtjni-maven-plugin</artifactId>
+          <version>1.18</version>
```
Comment (normanmaurer) [inline pom.xml:1829]: Why do we need this change ? I think this is not really related and should be done as a separate PR.

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: resolver-dns-native-windows/src/main/native-package/vs2010.custom.props [hunk 14]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="4.0">
+  <ItemDefinitionGroup>
+      <ClCompile>
+          <TreatWarningAsError>true</TreatWarningAsError>
+      </ClCompile>
+  </ItemDefinitionGroup>
+</Project>
```
Comment (chrisvest) [inline resolver-dns-native-windows/src/main/native-package/vs2010.custom.props:8]: Do we need this file? Also, I'm apprehensive about treating warnings as errors, since it can break the build down the line as newer compilers add more warnings.

**Comment by normanmaurer** [comment]
@flx5 please ping us when the comments were addressed

File: resolver-dns-native-windows/pom.xml (hunk 3 above)
Comment (flx5) [reply resolver-dns-native-windows/pom.xml]: Replaced cmake with hawtjni

File: resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c [hunk 15]
```diff
@@ -0,0 +1,380 @@
...
+JNIEXPORT jint JNI_OnLoad(JavaVM* vm, void* reserved) {
+    JNIEnv* env = NULL;
+
+    if ((*vm)->GetEnv(vm, (void**)&env, NETTY_JNI_UTIL_JNI_VERSION) != JNI_OK) {
+        fprintf(stderr, "FATAL: JNI version mismatch\n");
+        fflush(stderr);
+        return JNI_ERR;
+    }
+
+    if (loadClass(env, "io/netty/resolver/dns/windows/NativeException", &nativeExceptionClass) != JNI_OK) {
```
[the remark this answers is not in the document]
Comment (flx5) [reply resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c]: Classes are now loaded with NETTY_JNI_UTIL_LOAD_CLASS.

File: resolver-dns-native-windows/src/main/c/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,18 @@
+﻿cmake_minimum_required (VERSION 3.8)
```
[the remark this answers is not in the document]
Comment (flx5) [reply resolver-dns-native-windows/src/main/c/CMakeLists.txt]: Sorry, forgot about this one. Removed.

File: resolver-dns-native-windows/src/main/java/io/netty/resolver/dns/windows/NetworkAdapter.java [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.resolver.dns.windows;
+
+import java.net.InetSocketAddress;
+import java.util.List;
+
+public final class NetworkAdapter {
```
[the remark this answers is not in the document]
Comment (flx5) [reply resolver-dns-native-windows/src/main/java/io/netty/resolver/dns/windows/NetworkAdapter.java]: Yes, it can be. Marked as package private.

File: resolver-dns-native-windows/src/main/java/io/netty/resolver/dns/windows/NativeException.java [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.resolver.dns.windows;
+
+public class NativeException extends RuntimeException {
```
[the remark this answers is not in the document]
Comment (flx5) [reply resolver-dns-native-windows/src/main/java/io/netty/resolver/dns/windows/NativeException.java]: Replaced with generic runtime exception. Or is there a better existing class?

File: resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c [hunk 19]
```diff
@@ -0,0 +1,380 @@
...
+    }
+
+    if (dwRetVal == NO_ERROR) {
+        *target = pAddresses;
+        return JNI_OK;
+    }
+
+    if (pAddresses != NULL) {
+        free(pAddresses);
+    }
```
[the remark this answers is not in the document]
Comment (flx5) [reply resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c:118]: That won't work. In the successful case the caller has to free the IP_ADAPTER_ADDRESSES structure. Otherwise the target Variable would point to already freed memory.

File: resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c [hunk 20]
```diff
@@ -0,0 +1,380 @@
...
+	}
+	default:
+		throwNativeException(env, "Unknown address family %d", sockAddr->sa_family);
+		return NULL;
+	}
+
+    return (*env)->NewObject(env, inetSocketAddressClass, inetSocketAddressCtor, host, port);
+}
+
+jobject dnsServersToJava(JNIEnv* env, IP_ADAPTER_ADDRESSES* adapter) {
```
[the remark this answers is not in the document]
Comment (flx5) [reply resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c:163]: I tried using arrays at first, but then I'd have to iterate the c structure twice. One time to determine the size and another time to do the actual assignments. Not sure which variant is better here.

File: resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.h (hunk 11 above)
Comment (flx5) [reply resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.h]: Removed.

File: pom.xml (hunk 13 above)
Comment (flx5) [reply pom.xml:1829]: You are right, this should be a separate PR.
The issue here is that the plugin requires additional configuration to support newer MSVC versions: https://github.com/fusesource/hawtjni/issues/76.

That configuration property was however only available with the newer hawtjni version, which conveniently was also renamed...

Old versions of visual studio don't seem to be available anymore from microsoft, so I could not compile and test this otherwise.

File: resolver-dns-native-windows/src/main/native-package/vs2010.custom.props (hunk 14 above)
Comment (flx5) [reply resolver-dns-native-windows/src/main/native-package/vs2010.custom.props]: No, this is not required. I don't typically write C, so I wanted to be sure that I don't introduce new issues. Thus I decided to treat warnings as errors, because otherwise hawtjni tends to hide them.

**Comment by flx5** [comment]
@normanmaurer @chrisvest Sorry, didn't get around to this for quite some time. Although we are using this project (and other open source projects) at work, I am not allowed to work on open source during work hours. 

So writing this PR on my own time which can therefore be a little bit slow going.

- Do you think it would make more sense to first introduce the hawtjni in another PR before this PR?
- Do you think it has a relevant performance / readability impact to use arrays instead of lists in the native code?

**Comment by normanmaurer** [comment]
> @normanmaurer @chrisvest Sorry, didn't get around to this for quite some time. Although we are using this project (and other open source projects) at work, I am not allowed to work on open source during work hours.
> 
> So writing this PR on my own time which can therefore be a little bit slow going.

@flx5 no worries... 

> 
> * Do you think it would make more sense to first introduce the hawtjni in another PR before this PR?

Yes please.....

> * Do you think it has a relevant performance / readability impact to use arrays instead of lists in the native code?

We usually prefer to return arrays.

**Comment by flx5** [comment]
Hi @normanmaurer,

as discussed the hawtjni update has been placed in #13128.

I've also replaced the linked list with arrays and fixed some build issues.

Could you please review these changes?

Thank you!

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c [hunk 21]
```diff
@@ -0,0 +1,478 @@
...
+        }
+
+		port = ntohs(addr_in->sin6_port);
+		host = (*env)->NewStringUTF(env, ip);
+		break;
+	}
+	default:
+		throwNativeException(env, "Unknown address family %d", sockAddr->sa_family);
+		return NULL;
+	}
```
[the remark this answers is not in the document]
Comment (flx5) [reply resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c:155]: Oops. Sorry wrong IDE configuration re: tabs vs spaces

File: resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c [hunk 22]
```diff
@@ -0,0 +1,478 @@
...
+#include <winsock2.h>
+#include <Ws2ipdef.h>
+#include <ws2tcpip.h>
+#include <iphlpapi.h>
+#include <stdbool.h>
+
+#pragma comment(lib, "IPHLPAPI.lib")
+#pragma comment(lib, "ws2_32.lib")
+
+#define NETTY_JNI_UTIL_JNI_VERSION JNI_VERSION_1_6
```
[the remark this answers is not in the document]
Comment (flx5) [reply resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c]: Yes, that is from netty_jni_util.h
That was a remainder from the start when I didn't know about the util lib.
Removed that here.

**Review by normanmaurer — APPROVED** [review]

**Comment by normanmaurer** [comment]
@flx5 seems like you didn't sign our ICLA yet ? https://netty.io/s/icla ?

**Comment by flx5** [comment]
@normanmaurer Sorry, did not see that. I've signed it now.

**Comment by normanmaurer** [comment]
@flx5 thanks so much ... this is amazing
